src/services/mapCode.ts

```diff
@@ -0,0 +1,365 @@
+                if (!isBlock(node) && !isSourceFile(node)) {
+                    return false;
+                }
+                for (const stmt of node.statements) {
+                    for (const node of nodes) {
+                        if (matchNode(node, stmt)) {
+                            return true;
+                        }
+                    }
+                }
+                return false;
+            });
+            if (scope) {
+                // We found a scope that contains a matching node.
+
+                // First we find the last match of the range we're replacing.
+                // That is, assuming we have the following code to start with:
+                //
+                // ```ts
+                // function foo() {
+                //     console.log("hello");
+                // }
+                // function bar() {
+                //     console.log("world");
+                // }
+                // function baz() {
+                //     console.log("!");
+                // }
+                // function other() {
+                //    console.log("Leave me alone!");
+                // }
+                // ```
+                //
+                // And we get a map request with the following code:
+                //
+                // ```ts
+                // function bar() {
+                //     console.log("world");
+                // }
+                // function baz() {
+                //     console.log("!");
+                // }
+                // function qux() {
+                //     console.log("?");
+                // }
+                // ```
+                //
+                // `bar` will be the beginning of our range, and we want to
+                // have `baz` in the original be the end of our replacement
+                // range. `other` should be left intact, while `qux` should be
+                // "injected" before it.
+                //
+                // We do this by scanning the scope for all nodes that have a
+                // match in the incoming map request code, then sorting and
+                // picking first and last by position, and that will be our
+                // range for replacement.
+                const matches = (scope as Block).statements.filter(stmt => nodes.some(node => matchNode(node, stmt)));
+                matches.sort((a, b) => a.pos - b.pos);
+                changeTracker.replaceNodeRangeWithNodes(file, matches[0], matches[matches.length - 1], nodes);
+                // はい、できました！
+                return;
+            }
+        }
+    }
+
+    // If we have no frame of reference, we'll just insert at the end of the
+    // nearest statements scope, or, if we can't find a block, at the top
+    // level of the source file.
+    let scopeStatements: NodeArray<Statement> = file.statements;
+    top: for (const locationGroup of focusLocations) {

```

Comment:
This requires focusLocations to be flattened. The fallback to original isn't needed because findAncestor will walk to the root, which is the SourceFile `original`. So I put a `!` but maybe `?? original` would be better
```suggestion
    const scopeStatements = focusLocations.find(loc => findAncestor(getNodeAtPosition(original, loc.start), isBlock))!.statements;
```